### ISSUE TITLE
fix: The patch makes tryCatch invalid

### ISSUE THREAD

**Issue by LinXunFeng** [shorebirdtech/shorebird#1986]
App ID: N/A

**Description**

In version `1.0`, the `tryCatch` in the following code will become invalid after the patch is applied.

```dart
List<String> list = ['a', 'b'];
bool _isExist(String type) {
  try {
    final model = list.firstWhere(
      (element) => element.type == type,
    );
    return true;
  } catch (e) {
    return false;
  }
}

_isExist('c');
```

```dart
String value = '';
...
try {
  return json.decode(value);
} catch (e) {
  return null;
}
```
I'm glad this issue has been fixed in version `1.0.1`, but I found that the following code still makes `tryCatch` invalid, even in the latest version `1.0.4`.

```dart
class HexColor extends Color {
  static int _getColorFromHex(String hexColor) {
    try {
      hexColor = hexColor.toUpperCase().replaceAll("#", "");
      if (hexColor.length == 6) {
        hexColor = "FF$hexColor";
      }
      return int.parse(hexColor, radix: 16);
    } catch (e) {
      return 0x00000000;
    }
  }

  HexColor(final String hexColor) : super(_getColorFromHex(hexColor));
}

HexColor('');
```

The strange thing is that I new a demo and used the above code. I wanted to reproduce this problem but couldn't. It can only be reproduced in our commercial project.

This issue prevents our page from displaying properly, and it only appears on iOS.

**Expected Behavior**

The page can be displayed properly.

**Additional Context**

```shell
shorebird --version
Shorebird 1.0.4 • git@github.com:shorebirdtech/shorebird.git
Flutter 3.19.6 • revision aee5222c0d52012f089564a8395826d56e7c2fe8
Engine • revision e4434d7a212b29a8ee8ffaf8e2c453c753904933
```

**Comment by bryanoltman** [comment]
This is concerning. Without being able to reproduce the issue, it may be difficult for us to find the cause of the problem, but we will certainly try. Some questions for you:

1. When you say that the try/catch is "invalid", what exactly do you mean?
2. "This issue prevents our page from displaying properly" How? What do you expect to happen? What happens instead?
3. Would you be able to add logs to the code that has this issue to better understand what is going wrong? Without understanding what you're seeing on your end, it's tough to say exactly what these will look like, but being able to trace through the code and pinpoint exactly where it starts to do something we don't expect would be useful.

**Comment by bryanoltman** [comment]
Reopening, closed by accident

**Comment by eseidel** [comment]
@felangel and I did have to make a change to the "optimized try catch" logic inside the Dart VM.  The compiler typically optimizes away certain local variables in making the "catch" code.  It does so by holding onto direct references into the "Object Pool" which we are (later) sorting.  We were pretty sure we fixed the one case (in 1.0.1), but it's possible there is a second case?

Would *love* to work with you to fix.  I'm not sure how to reproduce from the code examples you're offering above though?

**Comment by LinXunFeng** [comment]
For demonstration, I adjusted the above `HexColor` and the changes are as follows (not important and can be ignored)
```diff
class HexColor extends Color {
  static int _getColorFromHex(String hexColor) {
    try {
      hexColor = hexColor.toUpperCase().replaceAll("#", "");
      if (hexColor.length == 6) {
        hexColor = "FF$hexColor";
      }
      return int.parse(hexColor, radix: 16);
    } catch (e) {
-      return 0x00000000;
+      return 0xFF000000;
    }
  }

  HexColor(final String hexColor) : super(_getColorFromHex(hexColor));
}
```

I added a new test page in our project, the code is as follows

```dart
Scaffold(
  backgroundColor: Colors.white,
  appBar: AppBar(
    backgroundColor: Color(0XFFFF8000),
    title: Text(
      'About us',
      style: TextStyle(
        fontSize: 18,
        color: Colors.white,
      ),
    ),
    centerTitle: true,
  ),
  body: Center(
    child: Text(
      'Hello world',
      style: TextStyle(
        color: HexColor(''), // This code will cause an exception after patching
        fontSize: 20,
      ),
    ),
  ),
);
```
 <img src="https://github.com/shorebirdtech/shorebird/assets/19367531/53314c29-521b-437a-9046-0c441bbbbe40" width = "300" align=center />

There is no error output on the console at this time.

Adjust the title color and then create the patch.

```diff
Scaffold(
  backgroundColor: Colors.white,
  appBar: AppBar(
    backgroundColor: Color(0XFFFF8000),
    title: Text(
      'About us',
      style: TextStyle(
        fontSize: 18,
-        color: Colors.white,
+        color: Colors.purple,
      ),
    ),
    centerTitle: true,
  ),
  body: Center(
    child: Text(
      'Hello world',
      style: TextStyle(
        color: HexColor(''), // This code will cause an exception after patching
        fontSize: 20,
      ),
    ),
  ),
);
```

```
[updater::network] Sending patch check request: PatchCheckRequest { app_id: "6fc7bf36-32b9-42f8-b79d-b7c7037a3627", channel: "stable", release_version: "7.1.5+1", patch_number: None, platform: "ios", arch: "aarch64" }
[updater::updater] Patch 1 successfully installed.
[updater::c_api] Update result: Update installed
```

This `app_id` belongs to my test account, and the patch is still there. It may be useful to you.


After installing the patch, restart the app and enter the `About us` page again.

 <img src="https://github.com/shorebirdtech/shorebird/assets/19367531/2b72f6e6-2150-4d2a-b3bc-ad62b5609153" width = "300" align=center />


The `About us` page show grey screen. The following is the error log output by the console.

```shell
flutter: Another exception was thrown: Instance of 'DiagnosticsProperty<void>'
```

The error log this time is different from what I encountered before. Maybe it’s because I didn’t change the same page.

The error I encountered before was this 

```shell
FormatException: Invalid radix-16 number`.
```

It was caused by `int.parse(hexColor, radix: 16);`, so I said tryCatch became invalid.

In addition, this time, my home page also displays a gray screen again, with the following error:

```shell
flutter: Bad state: No element
flutter: #0      ListBase.firstWhere (dart:collection/list.dart:124)
flutter: #1      HomeLogicMain._isExistMainModelType
flutter: #2      HomeLogicMain._updateMainEntryByType
...
```

This is the same problem as before when using version `1.0`.

<img width="300" alt="image" src="https://github.com/shorebirdtech/shorebird/assets/19367531/405dd227-5e0d-4555-b10c-92c8825c55cb">

**Comment by eseidel** [comment]
Thank you, this is very helpful.  We will investigate.

**Comment by sakhnyuk** [comment]
Same issue here. Only on iOS.

In catch block I show toasts or sometimes popup modal with information. After I applied patch in some cases catch block was completely skipped.

**Comment by eseidel** [comment]
Thank you very much.  We just need to get a reproduction (which we can try to do) and then it should be easy to fix.

**Comment by fiercepark** [comment]
@eseidel If this issue is fixed later, will I still need to release a new app version? Or is it based on the new shorebird generated patch and run on the old app version?

**Comment by felangel** [comment]
> @eseidel If this issue is fixed later, will I still need to release a new app version? Or is it based on the new shorebird generated patch and run on the old app version?

You’ll need to re-release as we’ve made some modifications to the snapshot format as part of #1892

**Comment by felangel** [comment]
@LinXunFeng I just tried to reproduce with the following configuration:

```
shorebird doctor -v       
Shorebird 1.1.6 • git@github.com:shorebirdtech/shorebird.git
Flutter 3.22.0 • revision 6d047e401941206daf11740c523ffa3364a2385b
Engine • revision c600950be82dcdad6a46da46bcc6e73773ca3907
```


Sample (`main.dart`)
```diff
import 'package:flutter/material.dart';

class HexColor extends Color {
  static int _getColorFromHex(String hexColor) {
    try {
      hexColor = hexColor.toUpperCase().replaceAll("#", "");
      if (hexColor.length == 6) {
        hexColor = "FF$hexColor";
      }
      return int.parse(hexColor, radix: 16);
    } catch (e) {
+      return Colors.pink.value;
-      return Colors.black.value;
    }
  }

  HexColor(final String hexColor) : super(_getColorFromHex(hexColor));
}

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.pink),
        useMaterial3: true,
      ),
      home: Scaffold(
        backgroundColor: Colors.white,
        appBar: AppBar(
          backgroundColor: const Color(0XFFFF8000),
          title: const Text(
            'Home',
            style: TextStyle(
              fontSize: 18,
              color: Colors.white,
            ),
          ),
          centerTitle: true,
        ),
        body: Center(
          child: Text(
            'Hello world',
            style: TextStyle(
              color: HexColor(''),
              fontSize: 20,
            ),
          ),
        ),
      ),
    );
  }
}
```

I was able to release and patch with no issues and the text color in the patch was pink.

Are you still able to reproduce this using Flutter 3.22 and Shorebird CLI v1.0.6?

**Comment by LinXunFeng** [comment]
@felangel  Due to some third-party packages, I cannot upgrade to Flutter 3.22 for the time being, and I am still using 3.19.6 and Shorebird CLI v1.1.6.

```shell
❯ shorebird --version
Shorebird 1.1.6 • git@github.com:shorebirdtech/shorebird.git
Flutter 3.22.0 • revision 6d047e401941206daf11740c523ffa3364a2385b
Engine • revision c600950be82dcdad6a46da46bcc6e73773ca3907
```

The result of my verification is that the problem still exists, but I adjusted the color of the title in the AppBar instead of `_getColorFromHex`.


```diff
appBar: AppBar(
backgroundColor: const Color(0XFFFF8000),
title: const Text(
  'Home',
  style: TextStyle(
    fontSize: 18,
-    color: Colors.white,
+    color: Colors.red,
  ),
),
```

**Comment by eseidel** [comment]
Update: This bug is very concerning, but our current theory is that we solved this in 3.22.0.  When you do get a chance to upgrade to 3.22.0, please let us know if it still reproduces for you.  We would be happy to hop on a call with you to debug if necessary.  Speculatively closing for now, please re-open if you learn anything.  Thank you very much!

**Comment by eseidel** [comment]
We think we might still have this bug.  @erickzanardo is investigating a test case we're seeing fail in the dart vm now.

**Comment by eseidel** [comment]
We thought we fixed this in https://github.com/shorebirdtech/dart-sdk/pull/418/files#diff-e2185101e2ff523ff7bf038c5aafe61d9462c4f51f06a69a20d03a84937d3cdaR373

But I see now reading it we labeled our fix as a "hack", it's possible that something else is confusing the CatchEntryMove modifications we're making there.

**Comment by eseidel** [comment]
I wonder if this broke with class table sorting?  Since class table sorting could end up causing us to create more object pool entries?  Maybe we're still applying the CatchEntryMove modifications even when we've otherwise disabled object pool sorting in the class-table sort pass?

**Comment by eseidel** [comment]
Update: We've reproduced and reduced this bug.  I expect we will have a fix on monday.  Thank you very much for the report!

**Comment by eseidel** [comment]
Here is the repro we ended up with:
```dart
void main() {
  // For some reason @pragma('shorebird:never-link')
  // is not sufficient to cause the exception not to be caught.
  print("Main is changed, but not the function call.");
  _foo();
}

void _foo() {
  try {
    // Because main has changed, the exception is not caught.
    throw "Should be caught!";
  } catch (e) {}
}
```

We're not yet sure what is getting confused in the patching process, but there isn't a lot of code involved here, so we expect we'll have a fix on Monday.

**Comment by eseidel** [comment]
Felix and Erick and I worked on this today (and solved it).  The problem was a confusion between how direct branches (for library-local static methods) can work and how the exception handling code works.  The exception handling code was computing an invalid pc_offset (number of bytes from the start of the function's code block) which was causing the exception handler look up to fail for the frame, and thus for the exception to fall through to the previous frame.  Felix is working on landing the change and rolling the engine now.

**Comment by eseidel** [comment]
Leaving open until we actually ship (should be tomorrow).

**Comment by eseidel** [comment]
Fix is ready to go out in the morning.  Thanks to @LinXunFeng and @erickzanardo for all the hard work getting us a reduction we could work from!

**Comment by eseidel** [comment]
This is fixed in 1.1.11 released today.  The fix requires using Flutter 3.22.2 or later.
